File: client/src/models/shift.js
import PropTypes from 'prop-types';

import { propTypes as userPropTypes } from './user';

export const propTypes = PropTypes.shape({
  endTime: PropTypes.string,
  id: PropTypes.number,
  published: PropTypes.bool,
  startTime: PropTypes.string,
  user: userPropTypes,
});

File: engines/schedules/app/models/shift.rb
# frozen_string_literal: true

class Shift < ApplicationRecord
  belongs_to :user
  has_one :company, through: :user

  validates :end_time, presence: true
  validates :start_time, presence: true
  validates :user, presence: true
  validate :start_time_is_before_end_time

  scope :for_company, ->(company) { joins(user: :company).where(users: { companies: { id: company.id } }) }
  scope :after, ->(time) { where(arel_table[:start_time].gteq(time)) }
  scope :before, ->(time) { where(arel_table[:start_time].lteq(time)) }
  scope :published, -> { where(published: true) }

  private

  def start_time_is_before_end_time
    return if start_time.blank? || end_time.blank?
    return if start_time < end_time
    errors.add(:start_time, "must be before end time")
  end
end
